### DISCUSSION TITLE
Raw preview for QueryParams in the request log

### DISCUSSION THREAD

**Discussion by marszysz** [mockoon/mockoon#634] [Ideas] [answered]
Hello,
It would be very helpful if one could see the raw query string in the request log. For now, if the query contains a structure, the preview isn't helpful at all:
![obraz](https://user-images.githubusercontent.com/19807080/152541689-ae1da33f-5bd4-40ac-9187-5eec070ad108.png)
Ideally, I would love to see some link like the one `View full body in editor` that would allow me to inspect the whole raw query string. For now, there's no means for doing this.

**Comment by 255kb** [comment]
Hi, 
I never realized that qs objects were displayed like this.  This is definitely not by design. 
Could you please open an issue to track this? 
Thanks!

**Comment by timon** [comment]
I'm facing the same problem. For me it would be enough if query params are just listed as a collection of 'key=value' pairs, without hierarchical display or other fancy stuff.

I wanted to come up with a PR that addresses it, but couldn't decide on what is the appropriate way to do it.

So far I have discovered, that Query Params come from Log Entries which are Trans created by mockoon-server. Mockoon server uses express under the hood, and express defaults to qs library when parsing query string.

This snippet seems to do the trick:

```js
const qs = require('qs');
const parseSimpleParams = str => qs.parse(str, { depth: 0 });
parseSimpleParams('a=b&c[d]=e'); // => { 'a': 'b', 'c[d]': 'e' }
```

However, I'm not sure which approach to follow here:
- Redefine `query parser` option in the server to use this function 
- Leave server as it is, and reprocess query params (by stringifying and then parsing again before displaying) them

I would perfer the first approach, but I'm not sure if this could cause unexpected consequences for requests matching.

**Reply by 255kb** [reply]
@timon The query params are already parsed by Express using Node.js' [`querystring`](https://nodejs.org/api/querystring.html) (it's the default parser) and available in `request.query`. 
As, the `Transaction` created in commons-server is only used to be displayed in the "Logs" tab and in the JSON logs of the CLI, I think approach number 2 would be the quickest and most logical. 
It's just a matter of choosing which objects and arrays representation is better. 

I quickly tried to add a `JSON.stringify()` [here](https://github.com/mockoon/commons-server/blob/6eddfc7dcb73449df02a2a9353f48bf3a14ce3cd/src/libs/utils.ts#L120-L125), and I get something like this: 
![image](https://user-images.githubusercontent.com/7489814/152992522-f9fdb07e-d035-44d5-bdc6-6c0a27c973fd.png)
I have mixed feelings, as no JSON was sent in the first place. It may a bit confusing.

Having something more like this:
![image](https://user-images.githubusercontent.com/7489814/152993424-e570e529-0d6b-46fd-b94e-df39bb57ff51.png)
or this:
![image](https://user-images.githubusercontent.com/7489814/152993636-be2640a0-e0ee-4c8c-a2df-cf4ec8478f95.png)
would be less confusing. 
But it would require a bit more work to get it right.

What do you think?

**Reply by timon** [reply]
For the sake of templating or filtering, I would prefer the latter

**Comment by timon** [comment]
The [express.js documentation](https://expressjs.com/en/4x/api.html#app.set) explicitly states that 'extended' (aka 'qs') is used by default, not 'querystring':
![image](https://user-images.githubusercontent.com/27407/153000827-75ec3b24-86ca-401b-8dde-b9451e46e663.png)

What I would really like for logs view is not to deal with any assumption on the structure of query parameters, and just explicitly list them how they were transferred:

```
prop1: value1
prop2[test1]: value1
prop2[test2]: value2
```

The reason I'd like to see it is to check if there are also some weirdness from the request side, like sending the same parameter key twice (`key=value1&key=value2`) or using dot notation, e.g `field[weird.dots]=true`

**Reply by 255kb** [reply]
I guess we could revert to accessing the original URL before parsing to display all its parts. However, this will enable you to debug what entered the server but now how the server parsed it and used it (in the templating helpers, response rules, etc.). 
For example, you would see a duplicated key like you suggested `key=value1&key=value2`, translated in the UI to:
```
key: value1
key: value2
```

But this would still be valid and parsed as an array by `qs` and used as an array by Mockoon, which IMO is a more important information than the raw string that was received.

**Reply by 255kb** [reply]
Ah! You are right default query parser is `qs`, my bad! (Not sure it changes anything for our issue here :))

**Reply by timon** [reply]
> But this would still be valid and parsed as an array by qs and used as an array by Mockoon, which IMO is a more important information than the raw string that was received.

Perhaps there are two different scenarios here, checking what was actually sent to the (mock) server, and exploration for the puproses of mocking. I might be over-engineering here, but maybe introduce a radio button to select which display style is desired, and have both?
At some point I thought of adding another property to transaction's request to hold simplistic array of parameters from query string just not to interfere with rules/templating engine

**Reply by 255kb** [reply, marked as answer]
Yes true. Maybe the UI could show the raw and full query string and the parsed result, something a bit like this: 
![image](https://user-images.githubusercontent.com/7489814/153037330-4e0776bd-a31d-4162-9f26-98bee24720e7.png)
